File: leetcode/algorithms/48_rotate_image/main.py
from typing import List


class RotateImage:
    """
    # Transpose & Reverse
    #
    # Complexities:
    #   N - Length of `matrix`
    #   - Time Complexity: O(N²)
    #   - Space Complexity: O(1)
    """
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        length: int = len(matrix)

        # Reverse rows
        length: int = len(matrix)

        for row in range(length // 2):
            for col in range(length):
                temp: int = matrix[row][col]
                matrix[row][col] = matrix[length - 1 - row][col]
                matrix[length - 1 - row][col] = temp

        # Transpose diagonally
        for row in range(length):
            for col in range(row + 1, length):
                temp: int = matrix[row][col]
                matrix[row][col] = matrix[col][row]
                matrix[col][row] = temp


    # Solution
    """
    # Solution 1
    #
    # Transpose & Reverse
    #
    # Complexities:
    #   N - Length of `matrix`
    #   - Time Complexity: O(N²)
    #   - Space Complexity: O(1)
    """
    def solution1(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        n = len(matrix)
        
        for i in range(n):
            for j in range(i, n):
                matrix[i][j], matrix[j][i] = matrix[j][i], matrix[i][j]
        
        for i in range(n):
            matrix[i].reverse()

    """
    # Solution 2
    #
    # Layer-by-Layer Rotation
    #
    # Complexities:
    #   N - Length of `matrix`
    #   - Time Complexity: O(N²)
    #   - Space Complexity: O(1)
    """
    def solution2(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        n = len(matrix)
        left, right = 0, n - 1
        
        while left < right:
            for i in range(right - left):
                top, bottom = left, right
                
                temp = matrix[top][left + i]
                
                matrix[top][left + i] = matrix[bottom - i][left]
                
                matrix[bottom - i][left] = matrix[bottom][right - i]
                
                matrix[bottom][right - i] = matrix[top + i][right]
                
                matrix[top + i][right] = temp
            
            left += 1
            right -= 1

File: leetcode/algorithms/48_rotate_image/main.java
class RotateImage {
    /**
     * Transpose & Reverse
     *
     * Complexities:
     *   N - Length of `matrix`
     *   - Time Complexity: O(N²)
     *   - Space Complexity: O(1)
     */
    public void rotate(int[][] matrix) {
        int length = matrix.length;

        // Reverse rows
        for (int row = 0; row < length / 2; row++) {
            int[] temp = matrix[row];
            matrix[row] = matrix[length - 1 - row];
            matrix[length - 1 - row] = temp;
        }

        // Transpose diagonally
        for (int row = 0; row < length; row++) {
            for (int col = row + 1; col < length; col++) {
                int temp = matrix[row][col];
                matrix[row][col] = matrix[col][row];
                matrix[col][row] = temp;
            }
        }
    }


    // Solution
    /**
     * Transpose & Reverse
     *
     * Complexities:
     *   N - Length of `matrix`
     *   - Time Complexity: O(N²)
     *   - Space Complexity: O(1)
     */
    public void solution(int[][] matrix) {
        int n = matrix.length;

        for (int i = 0; i < n; i++) {
            for (int j = i; j < n; j++) {
                int temp = matrix[i][j];
                matrix[i][j] = matrix[j][i];
                matrix[j][i] = temp;
            }
        }

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n / 2; j++) {
                int temp = matrix[i][j];
                matrix[i][j] = matrix[i][n - 1 - j];
                matrix[i][n - 1 - j] = temp;
            }
        }
    }
}
